text/0000-bind-helper.md

```diff
@@ -0,0 +1,124 @@
+
+Instead of this confusing and overloaded behavior, a new `bind` helper would be introduced to do partial application and context binding (with no need for build time private APIs).
+
+## Detailed design
+
+The `bind` helper will take in a function and then the set of arguments that will be partially applied to the function.
+As an optional named argument the user can pass in a `this` property that will set the `this` context this named argument would default to `undefined`.
+
+The behavior of this helper will be simple and use JavaScript's `function.prototype.bind`.
+
+Here are some examples of the `bind` helper and the equivalent JS:
+
+### Simple Case On Argument Curry
+
+```hbs
+{{bind this.log 1}}
+```
+
+```js
+this.log.bind(null, 1);
+```
+
+### Multiple Argument Partial Application
+
+```hbs
+{{bind this.add 1 2}}
+```
+
+```js
+this.add.bind(null, 1, 2);
+```
+
+### Setting the `this` context
+
+```hbs
+{{{bind this.session.logout context=this.anotherSession}}}
+```
+
+```js
+this.session.logout.bind(this.session)
+```
+
+### Setting the `this` context and arguments
+
+```hbs
+{{bind this.car.drive 1000 'mile' context=this.honda}}
+```
+
+```js
+this.car.logout.bind(this.car, 1000, 'mile')
+```
+
+### Comparison to Action Helper/Modifier
+
+```hbs
+{{!-- Actions --}}
+<button {{action "increment"}}>Click</button>
+<button {{action this.increment}}>Click</button>
+<button onclick={{action "increment"}}>Click</button>
+<button onclick={{action this.increment}}>Click</button>
+<button {{action (action "increment")}}>Click</button>
+<button {{action (action this.increment)}}>Click</button>
+
+<button onclick={{bind this.increment context=this}}>Click</button>
+<button {{on "click" (bind this.increment context=this)}}>Click</button>
+```
+
+### With `mut`
+
+```hbs
+{{!-- Actions --}}
+<button {{action (mut showModal) true}}>Click</button>
+<button onclick={{action (mut showModal) true}}>Click</button>
+<button {{action (action (mut showModal) true)}}>Click</button>
+
+<button onclick={{bind (mut showModal) true}}>Click</button>
+<button {{on "click" (bind (mut showModal) true)}}>Click</button>
+```
+
+Because `function.prototype.bind` always binds the `this` context the bind helper would require functions to be already bound or the `this` named argument would need to be explicitly set.
+
+## How we teach this
+
+For guides we would switch to recommending the `bind` helper to pass functions into components args and modifiers.
+In guides we would no longer recommend using the `action` helper based on the reasons listed in motivations.
+
+## Drawbacks
+
+Since this recommended design sets the `this` context to `null` using `function.prototype.bind` this could be confusing to junior developers if the function was not already bound (using something like an `@bound` decorator, JS binding, arrow functions, or the `bind` helper earlier in the stack).
```

Comment:
s/junior developers/new ember developers. IMO this context binding seems like a basic thing a framework should provide when template/JS class are already closely coupled. it's confusing if it's not out-of-box behavior.

Reply:
:+1: to s/junior developers/new ember developers/

While it might look like a good idea to have automagic context binding, this behaviour is closes to how JS and other libs work. Reducing the difference between JS and HBS is a great goal, IMHO.

Reply:
When calling an action on a component don't you almost always want `this` to be the component instance? would an `@action` called with `bind` be automatically have the `this` context set to the component instance?

Reply:
Correct, `@action` provides context binding. The magic that we’re discussing that is problematic is the fact that a _helper_ or _modifier_ can access this context implicitly.

Reply:
In which case the "how we teach this" section is misleading:

> For guides we would switch to recommending the bind helper to pass functions into components args and modifiers. In guides we would no longer recommend using the action helper based on the reasons listed in motivations.

This suggests you'd always want to use `bind` over `action` but I'd think that bind is only useful for the cases where you'd want a different context, which is (in my experience) never been a problem I've faced. Not to say that it's not a problem, but `targetObject` was how this was done in the past and that wasn't very pleasant.

Reply:
I think the missing piece is that we would recommend _against_ the `{{action}}` helper, and would recommend using the `@action` decorator for all actions. The `{{action}}` helper would likely be deprecated at some point in the future.

Reply:
For clarification (and something I will add to the RFC) is that we should recommend `bind` over `action` for many reasons but mainly because of the implicit `target=this` which is applied at build time.

For guides we would recommend `bind` without any `context=` because as @pzuraq says we'd recommend always using JS to bind context using the `@action` decorator or some other mechanism (like an `@autobind` decorator or even a language feature in the future)

Reply:
So for a button event listener to work, a dev would need to use {{bind}} in the template and @action in the JS. That seems like a weird thing to remember, IMO. @autobind or @bindcontext in the JS seems a little better IMO, but I know that’s a big departure from the current “action” terminology.

Reply:
So the way this will work would be that you'd decorate a method with `@action` and then call it with `{{bind}}` like:

```
@action
function increment(amount) {
  this.value = this.value + amount;
}

<button onclick={{bind this.increment 5}} />
```

The `@action` decorator in this case would be responsible for setting the correct context (e.g. setting the `this` to the component). Is that right?

Reply:
@jonnii That's my understanding, yes. Note also that with the RFC for the `{{on}}` modifier (#471)  you would now write `<button {{on "click" (bind this.increment 5)}} />`. v0.8 of the `ember-on-modifier` addon currently also allows you to omit the `bind` and just write `<button {{on "click" this.increment 5}} />`, but that isn't currently in the RFC.

Reply:
@mehulkar we already have to remember to declare actions inside `actions`. Using `@action` is a 1-1, so it would not be a problem, I guess.

Reply:
@rtablada, A little confused, what is the purpose of the default `target=this` being applied at build time? 

@viniciussbs Yeah, actions were a weird idiosyncrasy of the framework. bringing in plain JS bind, but keeping the idiosyncrasy makes things more weird for me, and just seems like an unnecessary hat tip to backwards compatibility. Wouldn’t it make more sense to just default to `context=this`?

Reply:
@mehulkar defaulting to `context=this` is more harmful than helpful. You might not be passing a function defined in the template context, but one coming from a service or a model, like `{{bind this.model.save}}`. That is the reason in `ember-bind-helper` we default to the everything but the last part of the call invocation (in the example `this.model`).

Reply:
When using `bind` with a context and an `@action` decorator which one wins?

Reply:
`@action`

Reply:
> That is the reason in ember-bind-helper we default to the everything but the last part of the call invocation

yeah this would be fine (and better) also. Defaulting to `null` is what I'm objecting to.